### ISSUE TITLE
clippy::let_and_return false positive

### ISSUE THREAD

**Issue by CreepySkeleton** [rust-lang/rust-clippy#4555]
### Summary
The `clippy::let_and_return` lint triggers from code generated by a macro-derive.

### Reproducer

```rust
use structopt::StructOpt;

#[derive(StructOpt)]
pub struct A {
    #[structopt(flatten)]
    pub b: B,
}

#[derive(StructOpt)]
pub struct B {}

fn main() {}
```

`structopt v0.3.2` does generate let_and_return in this case but I don't think this lint makes much sense when originated from a macro expansion.

# 
```sh
cargo clippy -V
clippy 0.0.212 (e3cb40e4 2019-06-25)
```

**Comment by llogiq** [comment]
That's strange. Gotta look into the code that structopt generates. Sometimes the expr spans aren't the same as their stmt spans.

**Comment by CreepySkeleton** [comment]
> Sometimes the expr spans aren't the same as their stmt spans.

Yes, structopt propagates source span info to the generated code. This particular piece of generated code has a span that points to `#[structopt(flatten)]` in source code.

**Comment by CreepySkeleton** [comment]
Relevant piece of expansion:
```rust
let app = <B>::augment_clap(app);
let app = if <B>::is_subcommand() {
    app.setting(::structopt::clap::AppSettings::SubcommandRequiredElseHelp)
} else {
    app
};
app
```

**Comment by CreepySkeleton** [comment]
[Full expansion](https://gist.github.com/CreepySkeleton/297b217066d98d03500c1f42613b0d1e)

**Comment by llogiq** [comment]
I think checking `if !in_external_macro(cx.sess(), retexpr.span);` in `returns.rs:210` should suffice. The return expression gets its own span from structopt, as far as I can see.

**Comment by CreepySkeleton** [comment]
> The return expression gets its own span from structopt, as far as I can see.

Not really.

https://github.com/TeXitoi/structopt/blob/80a55c0f35373779634cdb09fe556e19fedb0e77/structopt-derive/src/lib.rs#L103-L110

`kind.span()` here points to `flatten` in original source code

**Comment by CreepySkeleton** [comment]
@llogiq Yes, you're right
